pkg/controller/storagecluster/update.go

```diff
@@ -0,0 +1,604 @@
+	logrus.Debugf("Getting unavailable numbers")
+	nodeList := &v1.NodeList{}
+	err := c.client.List(context.TODO(), &client.ListOptions{}, nodeList)
+	if err != nil {
+		return -1, -1, fmt.Errorf("couldn't get list of nodes during rolling "+
+			"update of storage cluster  %#v: %v", cluster, err)
+	}
+
+	var numUnavailable, desiredNumberScheduled int
+	for _, node := range nodeList.Items {
+		wantToRun, _, _, err := c.nodeShouldRunStoragePod(&node, cluster)
+		if err != nil {
+			return -1, -1, err
+		}
+		if !wantToRun {
+			continue
+		}
+		desiredNumberScheduled++
+		storagePods, exists := nodeToStoragePods[node.Name]
+		if !exists {
+			numUnavailable++
+			continue
+		}
+		available := false
+		for _, pod := range storagePods {
+			// for the purposes of update we ensure that the pod is
+			// both available and not terminating
+			if podutil.IsPodReady(pod) && pod.DeletionTimestamp == nil {
+				available = true
+				break
+			}
+		}
+		if !available {
+			numUnavailable++
+		}
+	}
+
+	maxUnavailable, err := intstr.GetValueFromIntOrPercent(
+		cluster.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable,
+		desiredNumberScheduled,
+		true,
+	)
+	if err != nil {
+		return -1, -1, fmt.Errorf("invalid value for MaxUnavailable: %v", err)
+	}
+	logrus.Debugf("StorageCluster %s/%s, maxUnavailable: %d, numUnavailable: %d",
+		cluster.Namespace, cluster.Name, maxUnavailable, numUnavailable)
+	return maxUnavailable, numUnavailable, nil
+}
+
+func (c *Controller) cleanupHistory(
+	cluster *corev1alpha1.StorageCluster,
+	old []*apps.ControllerRevision,
+) error {
+	nodesToStoragePods, err := c.getNodeToStoragePods(cluster)
+	if err != nil {
+		return fmt.Errorf("couldn't get node to storage pod mapping for storage cluster %v: %v",
+			cluster.Name, err)
+	}
+
+	toKeep := int(*cluster.Spec.RevisionHistoryLimit)
+	toKill := len(old) - toKeep
+	if toKill <= 0 {
+		return nil
+	}
+
+	// Find all hashes of live pods
+	liveHashes := make(map[string]bool)
+	for _, pods := range nodesToStoragePods {
+		for _, pod := range pods {
+			if hash := pod.Labels[defaultStorageClusterUniqueLabelKey]; len(hash) > 0 {
+				liveHashes[hash] = true
+			}
+		}
+	}
+
+	// Find all live history with the above hashes
+	liveHistory := make(map[string]bool)
+	for _, history := range old {
+		if hash := history.Labels[defaultStorageClusterUniqueLabelKey]; liveHashes[hash] {
+			liveHistory[history.Name] = true
+		}
+	}
+
+	// Clean up old history from smallest to highest revision (from oldest to newest)
+	sort.Sort(historiesByRevision(old))
+	for _, history := range old {
+		if toKill <= 0 {
+			break
+		}
+		if liveHistory[history.Name] {
+			continue
+		}
+		// Clean up
+		err := c.client.Delete(context.TODO(), history)
+		if err != nil {
+			return err
+		}
+		toKill--
+	}
+
+	return nil
+}
+
+// isPodUpdated checks if pod contains label value that matches the hash
+func (c *Controller) isPodUpdated(cluster *corev1alpha1.StorageCluster, pod *v1.Pod, hash string) bool {
```

Comment:
The name for this is a little confusing. Can it be renamed to `isPodUpdateNeeded`?

Reply:
It is actually the reverse of that. We return true when we the pod is updated or at least selected fields have not changed (matched the current cluster spec). 
For clarity adding more comments in the function.